src/pandablocks_ioc/_hdf_ioc.py

```diff
@@ -21,15 +23,239 @@
+                "captured, disabling Capture."
+            )
+            self.status_message_setter("Requested number of frames captured")
+            self.put_data_to_file(EndData(self.number_of_received_rows, EndReason.OK))
+            self.finish_capturing = True
+
+    def _capture_forever(self, data: FrameData):
+        self.put_data_to_file(data)
+        self.number_of_received_rows += len(data.data)
+        self.number_received_setter(self.number_of_received_rows)
+
+    def _capture_last_n(self, data: FrameData):
+        """
+        Append every FrameData to a buffer until the number of rows equals
+        `:NumCapture`. Then rewrite the data circularly.
+
+        Only write the data once PCAP is received.
+        """
+        self.circular_buffer.append(data)
+        self.number_of_received_rows += len(data.data)
+
+        if self.number_of_received_rows > self.number_of_rows_to_capture:
+            self.status_message_setter(
+                "NumCapture received, rewriting first frames received"
+            )
+
+        else:
+            self.status_message_setter("Filling buffer to NumReceived")
+
+        while self.number_of_received_rows > self.number_of_rows_to_capture:
+            first_frame_data = self.circular_buffer.popleft()
+            first_frame_data_length = len(first_frame_data.data)
+
+            if first_frame_data_length >= self.number_of_rows_to_capture:
+                # More data than we want to capture, all in a single FrameData
+                # We can just slice with the NumCapture since this has to be the only FrameData
+                # in the buffer at this point
+                assert len(self.circular_buffer) == 0
+                first_frame_data.data = first_frame_data.data[-self.number_of_rows_to_capture:]
+                self.circular_buffer.appendleft(first_frame_data)
+                self.number_of_received_rows = self.number_of_rows_to_capture
+            elif (
+                first_frame_data_length > self.number_of_received_rows - self.number_of_rows_to_capture
+            ):
+                # We can slice from the beginning of the FrameData to have the desired number of rows
+                indices_to_discard = (
+                    self.number_of_received_rows - self.number_of_rows_to_capture
+                )
+                first_frame_data.data = first_frame_data.data[indices_to_discard:]
+                self.circular_buffer.appendleft(first_frame_data)
+                self.number_of_received_rows -= indices_to_discard
+                assert self.number_of_received_rows == self.number_of_rows_to_capture
+            else:
+                # If we remove the enire first frame data then the buffer will still
+                # be too big, or it will be exactly the number of rows we want
+                self.number_of_received_rows -= first_frame_data_length
+
+        self.number_received_setter(self.number_of_received_rows)
```

Comment:
@dmgav 

This mode adds [this functionality](https://github.com/dmgav/PandABlocks-ioc-ringbuf/blob/ea6a3b5cd82172747776d89df8c6e7d7edcca9a0/pandablocks_ioc_ringbuf/_hdf_ioc.py#L244-L377)

Reply:
I am planning to test this mode with Panda in the lab, but this will probably happen in the middle of the next week.

Reply:
@coretl